app/src/utils/DatasetTypes.ts

```diff
@@ -0,0 +1,47 @@
+/* TODO: These are not yet comprehensive, final interfaces */
+
+export interface DatasetMetadata {
+  id: string;
+  name: string;
+  description: string;
+  fields: Field[];
+}
+
+export interface Field {
+  data_type: string;
+  name: string;
+  description: string;
+  origin_dataset: string;
+}
+
+// TODO: make typedef for valid data types instead of any.
+export type Row = Record<string, any>;
+
+export class Dataset {
+  readonly rows: Row[];
+  readonly metadata: DatasetMetadata;
+
+  constructor(rows: Row[], metadata: DatasetMetadata) {
+    this.rows = rows;
+    this.metadata = metadata;
+  }
+
+  getTableViewColumns() {
+    return this.metadata.fields.map((field: Field) => ({
+      Header: field.name,
+      accessor: field.name,
+    }));
+  }
+}
+
+export type AllMetadata = Record<string, DatasetMetadata>;
+
+export type LoadStatus = "unloaded" | "loading" | "loaded" | "error";
+
+export interface DatasetStore {
+  readonly loadDataset: (id: string) => void;
+  readonly getDatasetLoadStatus: (id: string) => LoadStatus;
+  readonly metadataLoadStatus: LoadStatus;
+  readonly metadata: AllMetadata;
```

Comment:
I'm not sure I understand why you're creating a type for AllMetadata but not Record<string, Dataset>
My instinct is to prefer having Record versions for both as you get more visibility into what the object is without having to look up AllMetadata definition

If you do feel there's some value in keeping the type, I would add a comment at definition explaining that the string is the dataset ID to be explicit

Reply:
I was using this data type all over the place and it was pretty verbose. Also, comparing to the Dataset type, it's more common to need a reference to the whole metadata map than it is for the datasets where you'd usually just care about a specific one - eg metadata is one API call, and the catalog expects the whole map. I also updated it to be a read-only type to prevent mutating it without react noticing, which made it more verbose.

As an example, these were what it looked like before and I found that kind of hard to parse:
```
useResourceCache<Readonly<Record<string, Dataset>>>()
Promise<Readonly<Record<string, Dataset>>>
```

I changed the name to MetadataMap and a comment for the structure to make it clearer